File: SSNodeJsChat-Android/app/src/main/java/in/newdevpoint/ssnodejschat/observer/WebSocketSingleton.java
package in.newdevpoint.ssnodejschat.observer;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

import in.newdevpoint.ssnodejschat.AppApplication;
import in.newdevpoint.ssnodejschat.utility.PreferenceUtils;
import in.newdevpoint.ssnodejschat.utility.UserDetails;
import in.newdevpoint.ssnodejschat.webService.APIClient;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;

// Uses the Subject interface to update all Observers

public class WebSocketSingleton extends WebSocketListener implements DownloadSubject {
    private static final int NORMAL_CLOSURE_STATUS = 10000;

    private static final String TAG = "DownloadUtility";
    //    public static HashMap<String, DownloadRequest> downloadList = new HashMap<>();
    private static WebSocketSingleton webSocketSingleTon;
    private final ArrayList<WebSocketObserver> webSocketObservers;
    private WebSocket webSocket;

    public WebSocketSingleton() {

        // Creates an ArrayList to hold all observers
        webSocketObservers = new ArrayList<>();
        startWebSocket();
    }

    public static WebSocketSingleton getInstant() {
        if (webSocketSingleTon == null) {
            webSocketSingleTon = new WebSocketSingleton();
        }
        return webSocketSingleTon;
    }

    @Override
    public void register(WebSocketObserver newWebSocketObserver) {
        int observerIndex = webSocketObservers.indexOf(newWebSocketObserver);
        if (observerIndex == -1) {
            // Adds a new observer to the ArrayList
            webSocketObservers.add(newWebSocketObserver);
        } else {
            Log.d(TAG, "Subscriber is already registered");
        }


    }

    @Override
    public void unregister(WebSocketObserver deleteWebSocketObserver) {

        // Get the index of the observer to delete

        int observerIndex = webSocketObservers.indexOf(deleteWebSocketObserver);

        // Print out message (Have to increment index to match)

        System.out.println("Observer " + (observerIndex + 1) + " deleted");

        // Removes observer from the ArrayList
        if (observerIndex != -1)
            webSocketObservers.remove(observerIndex);

    }

    @Override
    public void notifyObserver(String response) {

        try {
            JSONObject jsonObject = new JSONObject(response);

            String responseType = jsonObject.getString("type");
            String message = jsonObject.getString("message");
            int statusCode = jsonObject.getInt("statusCode");

            // Cycle through all observers and notifies them of
            // price changes

            for (WebSocketObserver webSocketObserver : webSocketObservers) {
                Log.d(TAG, "notifyObserver: " + (webSocketObserver.getActivityName()));
                ResponseType[] registeredFor = webSocketObserver.registerFor();
                for (ResponseType element : registeredFor) {
                    if (element.equalsTo(responseType)) {
                        webSocketObserver.onWebSocketResponse(response, responseType, statusCode, message);
                        break;
                    }
                }


            }
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }

    private void startWebSocket() {

        // WebSocket
        Request request = new Request.Builder().url(APIClient.BASE_URL_WEB_SOCKET).build();

        OkHttpClient okHttpClient = new OkHttpClient();
        webSocket = okHttpClient.newWebSocket(request, this);
        okHttpClient.dispatcher().executorService().shutdown();
    }

    private void reconnectWebSocket() {

        // WebSocket
        Request request = new Request.Builder().url(APIClient.BASE_URL_WEB_SOCKET).build();

        OkHttpClient okHttpClient = new OkHttpClient();
        webSocket = okHttpClient.newWebSocket(request, this);
//        okHttpClient.dispatcher().executorService().shutdown();
    }


    public void sendMessage(JSONObject command) {
        webSocket.send(command.toString());
    }


    @Override
    public void onOpen(WebSocket webSocket, Response response) {
        Log.d(TAG, "WebSocket connect stable");
        joinCommand();
    }

    private void joinCommand() {
        if (PreferenceUtils.isUserLogin(AppApplication.applicationContext)) {
            UserDetails.myDetail = PreferenceUtils.getRegisterUser(AppApplication.applicationContext);

            JSONObject jsonObject = new JSONObject();
            try {

                jsonObject.put("user_id", PreferenceUtils.getRegisterUser(AppApplication.applicationContext).getId());
                jsonObject.put("type", "create");
                jsonObject.put(APIClient.KeyConstant.REQUEST_TYPE_KEY, APIClient.KeyConstant.REQUEST_TYPE_CREATE_CONNECTION);

                WebSocketSingleton.getInstant().sendMessage(jsonObject);

            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onMessage(WebSocket webSocket, final String text) {
        System.out.println("received message: " + text);
        notifyObserver(text);
    }


    @Override
    public void onClosing(WebSocket webSocket, int code, String reason) {
        System.out.println("onClosing: " + code + " / " + reason);
        closeConnection(webSocket);
    }

    private void closeConnection(WebSocket webSocket) {
        if (webSocket != null)
            webSocket.close(NORMAL_CLOSURE_STATUS, null);
    }

    @Override
    public void onFailure(WebSocket webSocket, Throwable t, Response response) {
        System.out.println("onFailure: " + t.getMessage());
//        closeConnection(webSocket);

        new Handler(Looper.getMainLooper()).postDelayed(this::reconnectWebSocket, 3000);
    }
}

File: SSNodeJsChat-Android-Kotlin/app/src/main/java/in/newdevpoint/ssnodejschat/observer/WebSocketSingleton.kt
package `in`.newdevpoint.ssnodejschat.observer

import `in`.newdevpoint.ssnodejschat.AppApplication
import `in`.newdevpoint.ssnodejschat.utility.PreferenceUtils
import `in`.newdevpoint.ssnodejschat.utility.UserDetails
import `in`.newdevpoint.ssnodejschat.webService.APIClient
import `in`.newdevpoint.ssnodejschat.webService.APIClient.KeyConstant
import android.os.Handler
import android.os.Looper
import android.util.Log
import okhttp3.*
import org.json.JSONException
import org.json.JSONObject
import java.util.*

// Uses the Subject interface to update all Observers
class WebSocketSingleton : WebSocketListener(), WebSocketSubject {
    private val webSocketObservers: ArrayList<WebSocketObserver>
    private var webSocket: WebSocket? = null
    override fun register(newWebSocketObserver: WebSocketObserver) {
        val observerIndex = webSocketObservers.indexOf(newWebSocketObserver)
        if (observerIndex == -1) {
            // Adds a new observer to the ArrayList
            webSocketObservers.add(newWebSocketObserver)
        } else {
            Log.d(TAG, "Subscriber is already registered")
        }
    }

    override fun unregister(deleteWebSocketObserver: WebSocketObserver) {

        // Get the index of the observer to delete
        val observerIndex = webSocketObservers.indexOf(deleteWebSocketObserver)

        // Print out message (Have to increment index to match)
        println("Observer " + (observerIndex + 1) + " deleted")

        // Removes observer from the ArrayList
        if (observerIndex != -1) webSocketObservers.removeAt(observerIndex)
    }

    override fun notifyObserver(response: String) {
        try {
            val jsonObject = JSONObject(response)
            val responseType = jsonObject.getString("type")
            val message = jsonObject.getString("message")
            val statusCode = jsonObject.getInt("statusCode")

            // Cycle through all observers and notifies them of
            // price changes
            for (webSocketObserver in webSocketObservers) {
                Log.d(TAG, "notifyObserver: " + webSocketObserver.activityName)
                val registeredFor = webSocketObserver.registerFor()
                for (element in registeredFor!!) {
                    if (element!!.equalsTo(responseType)) {
                        webSocketObserver.onWebSocketResponse(response, responseType, statusCode, message)
                        break
                    }
                }
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    private fun startWebSocket() {

        // WebSocket
        val request = Request.Builder().url(APIClient.BASE_URL_WEB_SOCKET).build()
        val okHttpClient = OkHttpClient()
        webSocket = okHttpClient.newWebSocket(request, this)
        okHttpClient.dispatcher().executorService().shutdown()
    }

    private fun reconnectWebSocket() {

        // WebSocket
        val request = Request.Builder().url(APIClient.BASE_URL_WEB_SOCKET).build()
        val okHttpClient = OkHttpClient()
        webSocket = okHttpClient.newWebSocket(request, this)
        //        okHttpClient.dispatcher().executorService().shutdown();
    }

    fun sendMessage(command: JSONObject) {
        webSocket!!.send(command.toString())
    }

    override fun onOpen(webSocket: WebSocket, response: Response) {
        Log.d(TAG, "WebSocket connect stable")
        joinCommand()
    }

    private fun joinCommand() {
        if (PreferenceUtils.Companion.isUserLogin(AppApplication.Companion.applicationContext)) {
            UserDetails.myDetail = PreferenceUtils.Companion.getRegisterUser(AppApplication.Companion.applicationContext)
            val jsonObject = JSONObject()
            try {
                jsonObject.put("user_id", PreferenceUtils.Companion.getRegisterUser(AppApplication.Companion.applicationContext).id)
                jsonObject.put("type", "create")
                jsonObject.put(KeyConstant.REQUEST_TYPE_KEY, KeyConstant.REQUEST_TYPE_CREATE_CONNECTION)
                getInstant()!!.sendMessage(jsonObject)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
    }

    override fun onMessage(webSocket: WebSocket, text: String) {
        println("received message: $text")
        notifyObserver(text)
    }

    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
        println("onClosing: $code / $reason")
        closeConnection(webSocket)
    }

    private fun closeConnection(webSocket: WebSocket?) {
        webSocket?.close(NORMAL_CLOSURE_STATUS, null)
    }

    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        println("onFailure: " + t.message)
        //        closeConnection(webSocket);
        Handler(Looper.getMainLooper()).postDelayed({ reconnectWebSocket() }, 3000)
    }

    companion object {
        private const val NORMAL_CLOSURE_STATUS = 10000
        private const val TAG = "DownloadUtility"

        //    public static HashMap<String, DownloadRequest> downloadList = new HashMap<>();
        private var webSocketSingleTon: WebSocketSingleton? = null
        fun getInstant(): WebSocketSingleton? {
            if (webSocketSingleTon == null) {
                webSocketSingleTon = WebSocketSingleton()
            }
            return webSocketSingleTon
        }
    }

    init {

        // Creates an ArrayList to hold all observers
        webSocketObservers = ArrayList()
        startWebSocket()
    }
}